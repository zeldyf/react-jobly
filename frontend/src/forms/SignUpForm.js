import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { Card, CardBody, Form, Label, Input, Button } from 'reactstrap';

/** User signup form.
 *
 * Shows form and manages update to state on changes.
 * On submission:
 * - calls signup function prop
 * - redirects to /companies route
 *
 * Routes -> SignupForm
 * Routed as /signup
 */

const SignUpForm = ({ signup }) => {
  const history = useHistory();
  const INITIAL_STATE = {
    username: '',
    password: '',
    firstName: '',
    lastName: '',
    email: ''
  };
  const [formData, setFormData] = useState(INITIAL_STATE);
  const [formErrors, setFormErrors] = useState([]);

  console.debug(
    'SignupForm',
    'signup=',
    typeof signup,
    'formData=',
    formData,
    'formErrors=',
    formErrors
  );

  /** Update form fields */
  const handleChange = evt => {
    const { name, value } = evt.target;
    setFormData(data => ({
      ...data,
      [name]: value
    }));
  };

  /** Handle form submit:
   *
   * Calls login func prop and, if successful, redirect to /companies.
   */
  const handleSubmit = async evt => {
    evt.preventDefault();
    try {
      let result = await signup(formData);
      // makes a POST request to Api.js and adds corresponding data to matching category in db.json
      if (result.success) {
        // imperatively redirect to correct page and refresh to see new data
        history.push('/companies');
      } else {
        setFormErrors(result.errors);
      }
    } catch (err) {
      console.log(err);
    }
  };

  return (
    <div className="SignUpForm col-md-5 offset-md-4 col-lg-4 offset-lg-4">
      <h1>Sign Up</h1>
      <Card>
        <CardBody>
          <Form className="SignUpForm-form" onSubmit={handleSubmit}>
            <Label htmlFor="username" className="SignUpForm-Label">
              Username
            </Label>
            <Input
              className="SignUpForm-Input"
              id="username"
              name="username"
              type="text"
              placeholder="Username"
              value={formData.username}
              onChange={handleChange}
              required
            ></Input>
            <Label htmlFor="password" className="SignUpForm-Label">
              Password
            </Label>
            <Input
              className="SignUpForm-Input"
              id="password"
              name="password"
              type="text"
              placeholder="Password"
              value={formData.password}
              onChange={handleChange}
              required
            ></Input>
            <Label htmlFor="firstName" className="SignUpForm-Label">
              First Name
            </Label>
            <Input
              className="SignUpForm-Input"
              id="firstName"
              name="firstName"
              type="text"
              placeholder="First Name"
              value={formData.firstName}
              onChange={handleChange}
              required
            ></Input>
            <Label htmlFor="lastName" className="SignUpForm-Label">
              Last Name
            </Label>
            <Input
              className="SignUpForm-Input"
              id="lastName"
              name="lastName"
              type="text"
              placeholder="Last Name"
              value={formData.lastName}
              onChange={handleChange}
              required
            ></Input>
            <Label htmlFor="email" className="SignUpForm-Label">
              Email
            </Label>
            <Input
              className="SignUpForm-Input"
              id="email"
              name="email"
              type="text"
              placeholder="Email"
              value={formData.email}
              onChange={handleChange}
              required
            ></Input>
            <span className="NewItemForm-formErrors">
              {formErrors ? <p>{formErrors}</p> : null}
            </span>
            <Button
              type="submit"
              className="btn btn-lg btn-block"
              color="primary"
            >
              Sign Up
            </Button>
          </Form>
        </CardBody>
      </Card>
    </div>
  );
};

export default SignUpForm;
