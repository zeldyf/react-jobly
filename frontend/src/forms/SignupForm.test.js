import React from 'react';
import { render } from '@testing-library/react';
import SignUpForm from './SignupForm';

test('renders SignupForm without crashing', () => {
  render(<SignUpForm />);
});


test('renders SignupForm correctly', () => {
  const { container } = render(<SignUpForm />);
  expect(container).toMatchSnapshot();
});
