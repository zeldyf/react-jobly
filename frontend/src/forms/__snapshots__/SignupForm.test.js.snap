// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SignupForm correctly 1`] = `
<div>
  <div
    class="SignUpForm col-md-5 offset-md-4 col-lg-4 offset-lg-4"
  >
    <h1>
      Sign Up
    </h1>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class="SignUpForm-form"
        >
          <label
            class="SignUpForm-Label"
            for="username"
          >
            Username
          </label>
          <input
            class="SignUpForm-Input form-control"
            id="username"
            name="username"
            placeholder="Username"
            required=""
            type="text"
            value=""
          />
          <label
            class="SignUpForm-Label"
            for="password"
          >
            Password
          </label>
          <input
            class="SignUpForm-Input form-control"
            id="password"
            name="password"
            placeholder="Password"
            required=""
            type="text"
            value=""
          />
          <label
            class="SignUpForm-Label"
            for="firstName"
          >
            First Name
          </label>
          <input
            class="SignUpForm-Input form-control"
            id="firstName"
            name="firstName"
            placeholder="First Name"
            required=""
            type="text"
            value=""
          />
          <label
            class="SignUpForm-Label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="SignUpForm-Input form-control"
            id="lastName"
            name="lastName"
            placeholder="Last Name"
            required=""
            type="text"
            value=""
          />
          <label
            class="SignUpForm-Label"
            for="email"
          >
            Email
          </label>
          <input
            class="SignUpForm-Input form-control"
            id="email"
            name="email"
            placeholder="Email"
            required=""
            type="text"
            value=""
          />
          <span
            class="NewItemForm-formErrors"
          >
            <p />
          </span>
          <button
            class="btn btn-lg btn-block btn btn-primary"
            type="submit"
          >
            Sign Up
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
